### PR TITLE
Implement agent removal logic and offering removal protection

### PR DESCRIPTION
**Pull request by nb-ohad**
- refactor finalizer addition and removal into utils
- add finalizer logic for ManageFusionOffering CR
- intreduce a new IsReadyToRemove offering callback
- add offering deletion logic when managed fusion config is marked for deletion
- lowercase error strings on some errors

### CODE REVIEW

File: controllers/datafoundation.go [hunk 1]
```diff
@@ -70,6 +70,10 @@ func DFAddToScheme(scheme *runtime.Scheme) {
 	utilruntime.Must(ocsv1.AddToScheme(scheme))
 }
 
+func dfIsReadyToBeRemoved(r *ManagedFusionOfferingReconciler, offering *v1alpha1.ManagedFusionOffering) (bool, error) {
```
Comment (leelavg) [inline controllers/datafoundation.go:73]: - Better have verb at the start in func sig, `dfIs` vs `isDf`

File: utils/utils.go [hunk 2]
```diff
@@ -51,8 +51,29 @@ func Remove(slice []string, s string) (result []string) {
...
+		time.Sleep(sleep)
+		log.Println("retrying after error:", err)
+	}
+	return fmt.Errorf("Failed after %d retries. Last error: %v", attempts, err)
+
+}
+
 // AddLabel add a label to a resource metadata
-func AddLabel(obj metav1.Object, key string, value string) {
+func AddLabel(obj client.Object, key string, value string) {
```
Comment (leelavg) [inline utils/utils.go:76]: - why `client.Object` if we are only going to touch `meta` on the object but not any `runtime`?

File: controllers/managedfusion_controller.go [hunk 3]
```diff
@@ -466,10 +484,10 @@ func (r *ManagedFusionReconciler) reconcileAlertmanagerSecret() error {
 			return err
 		}
 		if r.pagerDutyConfigData.ServiceKey == "" {
-			return fmt.Errorf("Agent PagerDuty configuration does not contain serviceKey entry")
+			return fmt.Errorf("agent PagerDuty configuration does not contain serviceKey entry")
```
Comment (leelavg) [inline controllers/managedfusion_controller.go:487]: - as we are touching these lines anyways, why not turn `fmt.Errorf` to `errors.New` if we aren't interpolating strings/errors?
- same for remaining

File: controllers/managedfusionoffering_controller.go [hunk 4]
```diff
@@ -247,6 +275,10 @@ func (r *ManagedFusionOfferingReconciler) getCSVByPrefix(name string) (*opv1a1.C
 
 // All the below functions are placeholder for offering plugin integration
 
+func pluginIsReadyToBeRemoved(reconciler *ManagedFusionOfferingReconciler, offering *v1alpha1.ManagedFusionOffering) (bool, error) {
```
Comment (leelavg) [inline controllers/managedfusionoffering_controller.go:278]: - better have verb at the start of func `pluginIs` vs `isPlugin`

File: controllers/managedfusion_controller.go [hunk 5]
```diff
@@ -297,26 +298,39 @@ func (r *ManagedFusionReconciler) initReconciler(ctx context.Context, req ctrl.R
...
 
 	} else if r.managedFusionSecret.UID != "" {
-		if !utils.Contains(r.managedFusionSecret.GetFinalizers(), managedFusionFinalizer) {
+		if added := utils.AddFinalizer(r.managedFusionSecret, managedFusionFinalizer); added {
 			r.Log.V(-1).Info(fmt.Sprintf("finalizer missing on the %s secret resource, adding...", managedFusionSecretName))
-			r.managedFusionSecret.SetFinalizers(append(r.managedFusionSecret.GetFinalizers(), managedFusionFinalizer))
 			if err := r.update(r.managedFusionSecret); err != nil {
 				return ctrl.Result{}, fmt.Errorf("failed to update %s secret with finalizer: %v", managedFusionSecretName, err)
 			}
+			r.Log.Info("Adding finallizer to %s secret completed successfuly", managedFusionSecretName)
```
Comment (leelavg) [inline controllers/managedfusion_controller.go:333]:
```suggestion
			r.Log.Info(fmt.Sprintf("Adding finallizer to %s secret completed successfuly", managedFusionSecretName))
```

File: controllers/datafoundation.go (hunk 1 above)
Comment (nb-ohad) [reply controllers/datafoundation.go:73]: the `df` prefix  is our internal marker for all offering-related callbacks, this is going to change when we separate the DF code into its own package

File: utils/utils.go (hunk 2 above)
Comment (nb-ohad) [reply utils/utils.go:76]: This is the new best practice for all `client` and `ctrlutil` functionality. They used to have segregation between functions that take `metav1.Object` and `runtime.Object` but then they introduced the `client.Object` which aggregates both. We are just following their footsteps

File: controllers/managedfusion_controller.go (hunk 3 above)
Comment (nb-ohad) [reply controllers/managedfusion_controller.go:487]: We can. But I find it easier to use `fmt.Errorf` for consistency and to ease changes were we will want to introduce variables into the message

File: controllers/managedfusionoffering_controller.go (hunk 4 above)
Comment (nb-ohad) [reply controllers/managedfusionoffering_controller.go:278]: the `plugin` prefix is our internal marker for all plugin infrastructure layer callbacks, this is going to change when we introduce an actual plugin abstraction.

File: controllers/managedfusion_controller.go (hunk 5 above)
Comment (nb-ohad) [reply controllers/managedfusion_controller.go]: Good catch! Updated

File: controllers/datafoundation.go (hunk 1 above)
Comment (leelavg) [reply controllers/datafoundation.go:73]: > this is going to change
- ack, the current functions always read as that something is ready to be removed 😅 
- fine for now, if this is slated for a refactor

File: controllers/managedfusion_controller.go (hunk 3 above)
Comment (leelavg) [reply controllers/managedfusion_controller.go:487]: - ack, I remember reading to use `errors.New` if there's no interpolation
- not that it gets in the same, so I don't feel any pressing need for the change

File: controllers/managedfusion_controller.go [hunk 6]
```diff
@@ -297,26 +298,39 @@ func (r *ManagedFusionReconciler) initReconciler(ctx context.Context, req ctrl.R
...
+			r.Log.Info("ManagedFusionOffering CRs found, issuing delete")
+			offering := v1alpha1.ManagedFusionOffering{}
+			if err := r.deleteAllOf(&offering); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to issue delete for one or more ManagedFusionOffering CRs, %v", err)
+			} else {
+				r.Log.Info("ManagedFusionOffering CRs delete issued, requeuing the event")
+				return ctrl.Result{RequeueAfter: 10}, nil
+			}
+		} else {
+			r.Log.Info("No offering found, removing finalizer from %s secret", managedFusionSecretName)
```
Comment (bindrad) [inline controllers/managedfusion_controller.go:314]: Use `fmt.Sprintf()` or pass key and value to `r.Log.Info()`
```suggestion
			r.Log.Info(fmt.Sprintf("No offering found, removing finalizer from %s secret", managedFusionSecretName))
```

File: controllers/managedfusion_controller.go [hunk 7]
```diff
@@ -297,26 +298,39 @@ func (r *ManagedFusionReconciler) initReconciler(ctx context.Context, req ctrl.R
...
+			if removed := utils.RemoveFinalizer(r.managedFusionSecret, managedFusionFinalizer); removed {
+				if err := r.update(r.managedFusionSecret); err != nil {
+					return ctrl.Result{}, fmt.Errorf("failed to remove finalizer from %s secret: %v", managedFusionSecretName, err)
+				}
+				r.Log.Info("Removing finallizer from %s secret completed successfuly", managedFusionSecretName)
```
Comment (bindrad) [inline controllers/managedfusion_controller.go:319]: Same as above, use `fmt.Sprintf`

File: controllers/managedfusion_controller.go [hunk 8]
```diff
@@ -297,26 +298,39 @@ func (r *ManagedFusionReconciler) initReconciler(ctx context.Context, req ctrl.R
...
-			r.Log.Info("finallizer removed successfully")
 
 			if err := r.initiateAgentUninstallation(); err != nil {
 				return ctrl.Result{}, fmt.Errorf("failed to initiate agent uninstallation %v", err)
 			}
 		}
 
 	} else if r.managedFusionSecret.UID != "" {
-		if !utils.Contains(r.managedFusionSecret.GetFinalizers(), managedFusionFinalizer) {
+		if added := utils.AddFinalizer(r.managedFusionSecret, managedFusionFinalizer); added {
```
Comment (bindrad) [inline controllers/managedfusion_controller.go:328]: Why can't we use the utils provided by controller-runtime instead of implementing our own?
https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.9.5/pkg/controller/controllerutil#AddFinalizer

File: controllers/managedfusion_controller.go [hunk 9]
```diff
@@ -297,26 +298,39 @@ func (r *ManagedFusionReconciler) initReconciler(ctx context.Context, req ctrl.R
...
+			offering := v1alpha1.ManagedFusionOffering{}
+			if err := r.deleteAllOf(&offering); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to issue delete for one or more ManagedFusionOffering CRs, %v", err)
+			} else {
+				r.Log.Info("ManagedFusionOffering CRs delete issued, requeuing the event")
+				return ctrl.Result{RequeueAfter: 10}, nil
+			}
+		} else {
+			r.Log.Info("No offering found, removing finalizer from %s secret", managedFusionSecretName)
+			if removed := utils.RemoveFinalizer(r.managedFusionSecret, managedFusionFinalizer); removed {
```
Comment (bindrad) [inline controllers/managedfusion_controller.go:315]: Why can't we use the utils provided by controller-runtime instead of implementing our own?
https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.9.5/pkg/controller/controllerutil#RemoveFinalizer

File: controllers/managedfusion_controller.go [hunk 10]
```diff
@@ -669,8 +687,8 @@ func (r *ManagedFusionReconciler) reconcileMonitoringResources() error {
 	r.Log.Info("reconciling monitoring resources")
 
 	podMonitorList := promv1.PodMonitorList{}
-	if err := r.list(&podMonitorList); err != nil {
-		return fmt.Errorf("Could not list pod monitors: %v", err)
+	if err := r.list(&podMonitorList, client.InNamespace(r.Namespace)); err != nil {
```
Comment (bindrad) [inline controllers/managedfusion_controller.go:690]: Why should we list podMonitors, serviceMonitors and promRules in single namespace only(the namespace where monitoring stack is deployed)?
We should add labels to the monitoring resources that are created by the offering as well.
Example: cephPromRules are created by ocs-operator so our prometheus should be able to read those rules and raise alerts, for this to happen, we should add label to the cephPromRules.

File: controllers/managedfusion_controller.go [hunk 11]
```diff
@@ -681,8 +699,8 @@ func (r *ManagedFusionReconciler) reconcileMonitoringResources() error {
 	}
 
 	serviceMonitorList := promv1.ServiceMonitorList{}
-	if err := r.list(&serviceMonitorList); err != nil {
-		return fmt.Errorf("Could not list service monitors: %v", err)
+	if err := r.list(&serviceMonitorList, client.InNamespace(r.Namespace)); err != nil {
```
Comment (bindrad) [inline controllers/managedfusion_controller.go:702]: Same as above

File: controllers/managedfusion_controller.go [hunk 12]
```diff
@@ -693,8 +711,8 @@ func (r *ManagedFusionReconciler) reconcileMonitoringResources() error {
 	}
 
 	promRuleList := promv1.PrometheusRuleList{}
-	if err := r.list(&promRuleList); err != nil {
-		return fmt.Errorf("Could not list prometheus rules: %v", err)
+	if err := r.list(&promRuleList, client.InNamespace(r.Namespace)); err != nil {
```
Comment (bindrad) [inline controllers/managedfusion_controller.go:714]: same as above

**Review by bindrad — CHANGES_REQUESTED** [review]

File: controllers/managedfusion_controller.go (hunk 8 above)
Comment (nb-ohad) [reply controllers/managedfusion_controller.go:328]: Because their finalizer utils do not return an indication if the new finalizer already exists or added.
We can change our utils to use a controllerutils implementation but that results in an unneeded iteration on the finalizer list which is unnecessary for such a simple operation

File: controllers/managedfusion_controller.go (hunk 9 above)
Comment (nb-ohad) [reply controllers/managedfusion_controller.go:315]: Because their finalizer utils do not return an indication if the finalizer was found and removed or not.
We can change our utils to use a controllerutils implementation but that results in an unneeded iteration on the finalizer list which is unnecessary for such a simple operation

File: controllers/managedfusion_controller.go (hunk 10 above)
Comment (nb-ohad) [reply controllers/managedfusion_controller.go:690]: Thinking about it again, it does not make sense to have this here at all. Each offering might have its own way of producing/creating the monitoring resources and might even reconcile back any changes to them.
For DF, OCS operator is adding he ability to ask it to add a desired label to the monitoring resources. This means the code above is not needed after we start deploying with ocs-operator 4.12. In the meantime, I think this code should move into `datafoundation.go` code, but that is for another PR.

File: controllers/managedfusion_controller.go (hunk 6 above)
Comment (nb-ohad) [reply controllers/managedfusion_controller.go]: done

File: controllers/managedfusion_controller.go (hunk 7 above)
Comment (nb-ohad) [reply controllers/managedfusion_controller.go]: done

File: controllers/managedfusion_controller.go (hunk 10 above)
Comment (nb-ohad) [reply controllers/managedfusion_controller.go]: I have reverted the changes to be moved to `datafoundation.go` at a future Pr

File: controllers/managedfusion_controller.go (hunk 11 above)
Comment (nb-ohad) [reply controllers/managedfusion_controller.go]: I have reverted the changes to be moved to `datafoundation.go` at a future Pr

File: controllers/managedfusion_controller.go (hunk 12 above)
Comment (nb-ohad) [reply controllers/managedfusion_controller.go]: I have reverted the changes to be moved to `datafoundation.go` at a future Pr

File: utils/utils.go (hunk 2 above)
Comment (nb-ohad) [reply utils/utils.go:76]: I don't remember how many operator SDK versions / kubebuilder versions ago they did the change where they refactor all of the APIs from `runtime.Object` and `meta.Object` to `client.Object` for their own controller utils. You can check the their history their release nodes for more information

File: cmd/awsDataGather/suite_test.go
Comment (ezio-auditore) [inline cmd/awsDataGather/suite_test.go:1]: This change is already in a different PR. Please remove unrelated changes in this PR

File: controllers/datafoundation.go
Comment (ezio-auditore) [inline controllers/datafoundation.go:1]: Lint changes should go into a seperate PR

**Review by ezio-auditore — CHANGES_REQUESTED** [review]
Please add managedfusionoffering CRD to the shim/crds directory for the tests to pass.
Also i see multiple files with lint changes. please keep lint changes in a seperate PR

**Comment by bindrad** [comment]
> Please add managedfusionoffering CRD to the shim/crds directory for the tests to pass.

We don't need to do this because it is taken care of from the config/crd folder. We need to add the agent package to the scheme in the test(https://github.com/red-hat-storage/managed-fusion-agent/blob/main/controllers/suite_test.go#L123).

File: controllers/datafoundation.go
Comment (nb-ohad) [reply controllers/datafoundation.go:1]: On a general note, there is no need to do a separate PR, we can have a separate commit. Saying that lint changes are automatic on my machine and every file which I open and save gets them. Without completely altering my setup it will be very hard and time-consuming which I don't think justifies the cost. Unless you have a suggestion for an easy way around it that does not include editing each file with a simple text editor, I would prefer not to do that. 

Also please note that these are not general lint changes, these are changes that are only applied to files touched by this PR

Comment (nb-ohad) [reply controllers/datafoundation.go:1]: A better approach would be to have a CI task that makes sure we are not contributing code that does not run through the full lint process. This will ensure that no one can contribute code that does not meet our code standards

Comment (nb-ohad) [reply controllers/datafoundation.go:1]: @ezio-auditore So I think we are in agreement that we need a future tuning to the CI, which will not be happing within this PR. Can this thread be resolved now?

Comment (ezio-auditore) [reply controllers/datafoundation.go:1]: Yes, but I am unable to resolve this. Something with the perms i guess. Please resolve this thread

File: cmd/awsDataGather/suite_test.go
Comment (ezio-auditore) [reply cmd/awsDataGather/suite_test.go:1]: This too can be resolved

**Review by bindrad — APPROVED** [review]
